### PR TITLE
Solution assignment 3

### PR DESCRIPTION
**Pull request by cba33**
I implemented the assignment using a graph traversal with branch pruning. Each letter is a node and the edges correspond to diagonal & straight neighbours of each letter in the input matrix. Finding words corresponds to a graph traversal. For efficiency, I prune the traversal branches when the word created so far does not correspond to a prefix in the dictionary. For the traversal, I chose depth-first search to arguably find full words faster than BFS. The worst-case complexity is O(rows * cols) , while the average-case complexity depends on the number of prefixes in the dictionary. In fact, how exactly should I reason about the average-case complexity?

### CODE REVIEW

File: Graph.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/*
+Each letter is a node. The graph has edges for each diagonal & straight traversal.
+Finding words means a graph traversal with word prunning. I chose DFS to
+potentially find complete words faster.
+ */
+public class Graph {
+
+    public final HashMap<Integer, Node> intToNode;
```
Comment (brzozan) [inline Graph.java:30]: It's a minor comment, but try to have types of fields (and return types of methods) declared using highest level interfaces (i.e. `Map<Integer, Node>`, `Map<Node, List<Node>` etc) rather than specific classes (`HashMap` etc). It makes future changes to the code easier and separates ideas (maps, lists) from their very specific implementations.

Also: could this integer simply be stored inside `Node` rather than in a separate map? It often is a good idea to keep data related to a single object inside it rather than in outside maps.

File: Graph.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    public final HashMap<Node, LinkedList<Node>> edges;
```
Comment (brzozan) [inline Graph.java:31]: Potential problem: `Node` does not have stable (as in: returning the same value for node objects mapped to the same integer identifier), so you rely on "pointer" to the object. It works here, but if someone changes the code a bit, they might get into some issues.
To avoid such problems one either defines `hashCode` and `equals` methods in the object used as key, or uses things like [AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md) to generate these automatically.
Alternatively you could use `Integer` (as defined in the map above) as a key.

That's from the code behavior perspective. From implementation perspective, if you need to use `Map<SomeObject, Something>`, it is usually a strong signal that `Something` should be simply a member of `SomeObject`.

File: Graph.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                int nodeKey = i * cols + j;
+                Node n = fetchOrCreateNode(nodeKey, input[i][j]);
+                LinkedList<Node> neigh = identifyNeighbors(input, i, j, rows, cols);
+                edges.put(n, neigh);
+            }
+        }
+    }
+
+    /* Build words using depth-first search and prune search using word prefixes. */
+    private void createWords(Node lastLetter, StringBuilder sb, HashSet visited, HashSet result) {
```
Comment (brzozan) [inline Graph.java:49]: Using containers without generic types specified is highly discouraged. Also: again, use interfaces rather than specific classes (`Set` instead of `HashSet`) unless you have a good reason to do otherwise.

File: Graph.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+        // debug
```
Comment (brzozan) [inline Graph.java:51]: Leaving commented out code is typically discouraged. Such code tends to rot and not be updated when class changes, so it stops working if uncommented.

File: Graph.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                    visited.remove(neigh);
+                    sb.deleteCharAt(sb.length()-1);
+                }
+            }
+        }
+    }
+
+    /* Initialize word creation from each node, if the dictioanry contains a prefix for that letter */
+    public HashSet createWords() {
+        HashSet visited = new HashSet();
```
Comment (brzozan) [inline Graph.java:78]: No type safety again... Please always use generic parameters to types. Java compiler probably gives you a warning about it. It is a good practice to always fix compiler warnings, even if they seem silly. Compilers are really good at pointing out potential problems.

File: Graph.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        return result;
+    }
+
+
+    private LinkedList<Node> identifyNeighbors(char[][] input, int i, int j, int rows, int cols) {
+        LinkedList<Node> neighbors = new LinkedList<>();
+        int pos;
+        char c;
+
+        if(i-1 >= 0) {
```
Comment (brzozan) [inline Graph.java:99]: Could this sequence of ifs be expressed using a loop? There seems to be a lot of code duplication in these.

File: GraphTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+}
+
+class GraphTest {
+
+    char[][] input = {{'A','A','R'}, {'T','C','D'}};
+    Graph g = new Graph(2, 3, input);
+
+    @org.junit.jupiter.api.Test
```
Comment (brzozan) [inline GraphTest.java:44]: Why not import the class and use just `@Test`?

File: GraphTest.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    static boolean isPrefix(String str) {
+        return prefixes.contains(str);
+    }
+
+}
+
+class GraphTest {
+
+    char[][] input = {{'A','A','R'}, {'T','C','D'}};
```
Comment (brzozan) [inline GraphTest.java:41]: private final.

File: GraphTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    Graph g = new Graph(2, 3, input);
```
Comment (brzozan) [inline GraphTest.java:42]: Typically tested object is created inside a method annotated with `@Before`.

Also: private, unless there is a good reason to make it visible to the whole package.

File: GraphTest.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    @org.junit.jupiter.api.Test
+    void createWords() {
+        HashSet wordsFound = g.createWords();
+        System.out.print(wordsFound);
+        if (wordsFound.size() != 3)
```
Comment (brzozan) [inline GraphTest.java:48]: Why not use [assertions](http://junit.sourceforge.net/javadoc/org/junit/Assert.html) instead of ifs?

Also: please avoid ifs without `{..}`. Brackets don't cost much, but make code more consistent and easier to read, as well as to avoid easy mistakes when modifying it later.

File: GraphTest.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        return words.contains(str);
+    }
+
+    static boolean isPrefix(String str) {
+        return prefixes.contains(str);
+    }
+
+}
+
+class GraphTest {
```
Comment (brzozan) [inline GraphTest.java:39]: No tests for edge cases like empty grid, grid with no words present in the dictionary etc...

File: Graph.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+
+/**
+ * Created by basescu on 06.06.17.
+ */
+
+class Node {
+    public int key;     // letter index in initial matrix: for M[i][j], we have pos = i * cols + j
```
Comment (brzozan) [inline Graph.java:10]: final! Especially if field is publicly accessible and thus could be modified by someone outside your control (imagine that this code is part of a bigger project, where more people can use it).
Same for value below.

**Review by brzozan — CHANGES_REQUESTED** [review]
Sorry for late review, but your solution was most complicated :)

Good, but Ive had a few comments.
